pangeo_forge_recipes/patterns.py

```diff
@@ -129,11 +129,29 @@ class FilePattern:
       list.
     :param combine_dims: A sequence of either concat or merge dimensions. The outer
       product of the keys is used to generate the full list of file paths.
+    :param fsspec_open_kwargs: Extra options for opening the inputs with fsspec.
+      May include ``block_size``, ``username``, ``password``, etc.
+    :param query_string_secrets: If provided, these key/value pairs are appended to
+      the query string of each ``file_pattern`` url at runtime.
+    :param is_opendap: If True, assume all input fnames represent opendap endpoints.
+      Cannot be used with caching.
     """
 
-    def __init__(self, format_function: Callable, *combine_dims: CombineDim):
+    def __init__(
+        self, format_function: Callable, *combine_dims: CombineDim, **kwargs,
+    ):
         self.format_function = format_function
         self.combine_dims = combine_dims
+        self.fsspec_open_kwargs = kwargs.pop("fsspec_open_kwargs", {})
+        self.query_string_secrets = kwargs.pop("query_string_secrets", {})
+        self.is_opendap = kwargs.pop("is_opendap", False)
```

Comment:
☑️ https://github.com/pangeo-forge/pangeo-forge-recipes/pull/167/commits/efbf3354bf3a05ee31800ed12c28c6aaeff58c1f

Oh! I hadn't realized that this was an allowable signature:

```python
Signature: func(positional_arg, *args, default_kwarg=None)
```

(I'd mistakenly thought *args needed to come after default kwargs.)

Agreed that this is *way* better.

I've set both the `fsspec_open_kwargs` and `query_string_secrets` defaults to an empty dict (rather than None), because there may be scenarios where we want to call, e.g., `fsspec_open_kwargs.update()` at runtime.

Reply:
> defaults to an empty dict (rather than None), because there may be scenarios where we want to call, e.g., `fsspec_open_kwargs.update()` at runtime.

This is actually a common python pitfall, as the following code will illustrate.

```python
class Foo
    def __init__(self, bar={}):
        self.bar = bar

f1 = Foo()
f2 = Foo()
f1.bar.update({'a': 1})
print(f2.bar) # what do you think it will be?
```

This is [solved in dataclasses](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) by using the `default_factory=dict`. If we don't go with dataclasses, you need to do something like
```python
if fsspec_open_kwargs is None:
    fsspec_open_kwargs = {}
```
inside `__init__`.

Reply:
🤯

Reply:
For a dataclass to support variable numbers of `*args` (i.e., `*combine_dims`), we need to set `@dataclass(init=False)` and override the default `__init__`: https://www.python.org/dev/peps/pep-0557/#custom-init-method

Expanding on your toy example, that would look like this, for our use case (with incorrect types, for illustration):
```python
from dataclasses import dataclass

@dataclass(init=False)
class FilePattern:
    format_function: str
    combine_dims: int
    bar: dict
    def __init__(self, format_function, *combine_dims, bar=None):
        self.format_function = format_function
        self.combine_dims = combine_dims
        self.bar = bar if bar else {}

f1 = FilePattern("func", 0, 1)
f2 = FilePattern("func", 0)
f1.bar.update({'a': 1})
print(
    f"f1: {vars(f1)}\n"
    f"f2: {vars(f2)}"
)
```
```
f1: {'args': (0, 1), 'bar': {'a': 1}}
f2: {'args': (0,), 'bar': {}}
```
This means we can't leverage `field=(default_factory=dict)` (which is passed to the default dataclass `__init__`) and instead need to fall back to what you've proposed for non-dataclasses, i.e. `self.bar = bar if bar else {}`.

The only problem with this is that the dataclass type hints don't make it into the `__init__` signature (unless we put them there, redundantly), and will therefore be absent from the auto-generated API documentation for this class. I checked this with `make html`, but you can see it more simply here
```python
FilePattern?
```
```python
Init signature: FilePattern(format_function, *combine_dims, bar=None)
Docstring:      FilePattern(format_function, *combine_dims, bar=None)
Type:           type
Subclasses:   
```
As such, I suggest we go with the more concise non-dataclass approach:

```python
class FilePattern:
    def __init__(self, format_function: str, *combine_dims: int, bar: dict=None):
        self.format_function = format_function
        self.combine_dims = combine_dims
        self.bar = bar if bar else {}

f1 = FilePattern("func", 0, 1)
f2 = FilePattern("func", 0)
f1.bar.update({'a': 1})
print(
    f"f1: {vars(f1)}\n"
    f"f2: {vars(f2)}"
)
```
```
f1: {'format_function': 'func', 'combine_dims': (0, 1), 'bar': {'a': 1}}
f2: {'format_function': 'func', 'combine_dims': (0,), 'bar': {}}
```
Which preserves type hints that are propagated to sphinx:
```
FilePattern?
```
```
Init signature: FilePattern(format_function: str, *combine_dims: int, bar: dict = None)
Docstring:      <no docstring>
Type:           type
Subclasses: 
```

Thoughts?

Reply:
> For a dataclass to support variable numbers of `*args` (i.e., `*combine_dims`), we need to set `@dataclass(init=False)` and override the default `__init__`

Gah I did not realize this. Maybe not worth it then.

Reply:
> As such, I suggest we go with the more concise non-dataclass approach:

:+1: